In [16]:
#lab9
import numpy
import scipy.special
import matplotlib.pyplot
%matplotlib inline

In [17]:
class neuralNetwork:
    
    def __init__(self, inputnodes, hiddennodes1, hiddennodes2, outputnodes, learningrate):
        self.inodes = inputnodes
        self.hnodes1 = hiddennodes1
        self.hnodes2 = hiddennodes2
        self.onodes = outputnodes
        
        self.wi1 = numpy.random.normal(0.0, pow(self.hnodes1, -0.5), (self.hnodes1, self.inodes))
        self.w12 = numpy.random.normal(0.0, pow(self.hnodes2, -0.5), (self.hnodes2, self.hnodes1))
        self.w2o = numpy.random.normal(0.0, pow(self.onodes, -0.5), (self.onodes, self.hnodes2))
        
        self.lr = learningrate
        
        self.activation_function = lambda x: scipy.special.expit(x)
        
        pass
    
    def train(self, inputs_list, targets_list):
        inputs = numpy.array(inputs_list, ndmin=2).T
        targets = numpy.array(targets_list, ndmin=2).T
        
        hidden_inputs_1 = numpy.dot(self.wi1, inputs)
        hidden_outputs_1 = self.activation_function(hidden_inputs_1)
        
        hidden_inputs_2 = numpy.dot(self.w12, hidden_outputs_1)
        hidden_outputs_2 = self.activation_function(hidden_inputs_2)
        
        final_inputs = numpy.dot(self.w2o, hidden_outputs_2)
        final_outputs = self.activation_function(final_inputs)
        
        output_errors = targets - final_outputs
        hidden_errors_2 = numpy.dot(self.w2o.T, output_errors)
        hidden_errors_1 = numpy.dot(self.w12.T, hidden_errors_2)
        
        self.w2o += self.lr * numpy.dot((output_errors * final_outputs * (1.0 - final_outputs)), numpy.transpose(hidden_outputs_2))
        
        self.w12 += self.lr * numpy.dot((hidden_errors_2 * hidden_outputs_2 * (1.0 - hidden_outputs_2)), numpy.transpose(hidden_outputs_1))
        
        self.wi1 += self.lr * numpy.dot((hidden_errors_1 * hidden_outputs_1 * (1.0 - hidden_outputs_1)), numpy.transpose(inputs))
        
        pass
    
    def query(self, inputs_list):
        inputs = numpy.array(inputs_list, ndmin=2).T
        
        hidden_inputs_1 = numpy.dot(self.wi1, inputs)
        hidden_outputs_1 = self.activation_function(hidden_inputs_1)
        
        hidden_inputs_2 = numpy.dot(self.w12, hidden_outputs_1)
        hidden_outputs_2 = self.activation_function(hidden_inputs_2)
        
        final_inputs = numpy.dot(self.w2o, hidden_outputs_2)
        final_outputs = self.activation_function(final_inputs)
        
        return final_outputs

In [37]:
#노드 수 설정
input_nodes = 784
hidden_nodes1 = 100
hidden_nodes2 = 100
output_nodes = 10

learning_rate = 0.1
n = neuralNetwork(input_nodes, hidden_nodes1, hidden_nodes2, output_nodes, learning_rate)

In [38]:
#mnist 학습 데이터인 csv 파일을 리스트로 불러오기
training_data_file = open("mnist_train.csv","r")
training_data_list = training_data_file.readlines()
training_data_file.close()

In [39]:
#신경망 학습시키기

epochs = 5

for e in range(epochs):
    #학습 데이터 모음 내의 모든 레코드 탐색
    for record in training_data_list:
        #레코드를 쉼표에 의해 분리
        all_values = record.split(',')
        #입력 값의 범위와 값 조정
        inputs = (numpy.asfarray(all_values[1:]) / 255.0 * 0.99) + 0.01
        #결과 값 생성 (실제 값인 0.99 외에는 모두 0.01)
        targets = numpy.zeros(output_nodes) + 0.01
        #all_values[0]은 이 레코드에 대한 결과 값
        targets[int(all_values[0])] = 0.99
        n.train(inputs, targets)
        pass
pass

In [40]:
#mnist 테스트 데이터인 csv 파일을 리스트로 불러오기
test_data_file = open("mnist_test.csv","r")
test_data_list = test_data_file.readlines()
test_data_file.close()

In [41]:
#신경망 테스트

#신경망의 성능의 지표가 되는 성적표를 아무 값도 가지지 않도록 초기화
scorecard = []

#테스트 데이터 모음 내의 모든 레코드 탐색
for record in test_data_list:
    #레코드를 쉼표에 의해 분리
    all_values = record.split(',')
    #정답은 첫 번째 값
    correct_label = int(all_values[0])
    #입력 값의 범위와 값 조정
    inputs = (numpy.asfarray(all_values[1:]) / 255.0 * 0.99) + 0.01
    #신경망에 질의
    outputs = n.query(inputs)
    #가장 높은 값의 인덱스는 레이블의 인덱스와 일치
    label = numpy.argmax(outputs)
    #정답 또는 오답을 리스트에 추가
    if(label == correct_label):
        #정답인 경우 성적표에 1을 더해 줌
        scorecard.append(1)
    else:
        #정답이 아닌 경우 성적표에 0을 더해줌
        scorecard.append(0)
        pass
    pass

In [42]:
#정답의 비율인 성적을 계산해 출력
scorecard_array = numpy.asarray(scorecard)
print("epochs = ", epochs,
      "hidden_nodes1 = ", hidden_nodes1, "hidden_nodes2", hidden_nodes2,
      "learning_rate = ", learning_rate,
      "performance = ", scorecard_array.sum() / scorecard_array.size)
print(" KIM Min Gyu")

epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9335
 KIM Min Gyu


In [15]:
print("lab9 김민규 201702986")

lab9 김민규 201702986


attempt1
epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9204
KIM Min Gyu

attempt2
epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9273
KIM Min Gyu

attempt3
epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9189
KIM Min Gyu

attempt4
epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9406
 KIM Min Gyu
 
attempt5
epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9335
 KIM Min Gyu

highest result
epochs =  5 hidden_nodes1 =  100 hidden_nodes2 100 learning_rate =  0.1 performance =  0.9406
 KIM Min Gyu